FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs

```diff
@@ -25,49 +25,36 @@ public static ElementNode Anonymize(this ElementNode node, AnonymizationFhirPath
             return node;
         }
 
-        // Remove null children of current node, and return true => current node is null
-        public static bool RemoveNullChildren(this ElementNode node)
+        public static void RemoveNullChildren(this ElementNode node)
         {
             if (node == null)
             {
-                return true;
+                return;
             }
 
-            var children = node.Children().CastElementNodes().ToList();
+            var children = node.Children().ToList();
             foreach (var child in children)
             {
-                // Remove child if it is null => return true
-                if (RemoveNullChildren(child))
+                var elementNodeChild = (ElementNode)child;
+
+                // Remove null children recursively
+                RemoveNullChildren(elementNodeChild);
+
+                if (ShouldRemoveNode(elementNodeChild))
                 {
-                    node.Remove(child);
+                    node.Remove(elementNodeChild);
                 }
             }
-
-            bool currentNodeIsEmpty = !node.Children().Any() && node.Value == null;
-            bool currentNodeIsFhirResource = node.IsFhirResource();
-            if (currentNodeIsEmpty && !currentNodeIsFhirResource)
-            {
-                return true;
-            }
-            else
-            {
-                return false;
-            }
         }
 
         public static void AddSecurityTag(this ElementNode node, ProcessResult result)
```

Comment:
I feel this function is not belong to this class @sowu880

Reply:
I feel ProcessResult doesn't belong to this node

Reply:
By offline discussion, it is moved to ResourceProcessor.